consensus/istanbul/backend/engine.go

```diff
@@ -175,20 +177,112 @@ func (sb *Backend) verifyCascadingFields(chain consensus.ChainReader, header *ty
 func (sb *Backend) VerifyHeaders(chain consensus.ChainReader, headers []*types.Header, seals []bool) (chan<- struct{}, <-chan error) {
 	abort := make(chan struct{})
 	results := make(chan error, len(headers))
-	go func() {
-		for i, header := range headers {
-			err := sb.verifyHeader(chain, header, headers[:i])
 
-			select {
-			case <-abort:
-				return
-			case results <- err:
+	// If we are in "lightest" mode, we do batch verification of all BLS signatures
+	// in the headers, so there is no need to iterate over the blocks
+	if chain.Config().FullHeaderChainAvailable {
+		go func() {
+			for i, header := range headers {
+				err := sb.verifyHeader(chain, header, headers[:i])
+
+				select {
+				case <-abort:
+					return
+				case results <- err:
+				}
 			}
-		}
-	}()
+		}()
+	} else {
+		go func() {
+			err := sb.batchVerifyHeaders(chain, headers)
+			// even though we get 1 return value from the batched verification
+			// method, we need to fill the whole channel of results with return values
+			for i := 0; i < len(headers); i++ {
+				select {
+				case <-abort:
+					return
+				case results <- err:
+				}
+			}
+		}()
+	}
 	return abort, results
 }
 
+// batchVerifyHeaders is called in ultralight mode and performs batch BLS signature
+// verification over all provided headers and their corresponding validator sets
+func (sb *Backend) batchVerifyHeaders(chain consensus.ChainReader, headers []*types.Header) error {
+	// TODO: Populating this array can easily be done in parallel since the order does not matter
+	var msgs []*bls_ffi.SignedBlockHeader
+	for i, header := range headers {
+		// skip the genesis block
+		if header.Number.Uint64() == 0 {
+			continue
+		}
+
+		extra, err := types.ExtractIstanbulExtra(header)
+		if err != nil {
+			return err
+		}
+
+		// Don't waste time checking blocks from the future (allow some minor clock skew)
+		if header.Time > uint64(now().Unix())+mobileAllowedClockSkew {
+			return consensus.ErrFutureBlock
+		}
+
+		// short-circuit if there was no signature found
+		if len(extra.AggregatedSeal.Signature) == 0 {
+			return errEmptyAggregatedSeal
+		}
+
+		// 1. Get the aggregate public key for this epoch
+
+		// 1.1. Get the snapshot
+		snap, err := sb.snapshot(chain, header.Number.Uint64()-1, header.ParentHash, headers[:i])
+		if err != nil {
+			return err
+		}
+
+		// 1.2 Get the subset of validator public keys that signed
+		// based on the sig's bitmap
+		asig := extra.AggregatedSeal
+		publicKeys, err := getAggregatedSealSigners(snap.ValSet, asig)
+		if err != nil {
+			return err
+		}
+
+		// 1.3. Aggregate them
+		apk, err := blscrypto.AggregatePublicKeys(publicKeys)
```

Comment:
Is this step expensive? Is there anything we can do to take advantage of the fact that the validator set changes slowly?